src/components/Chat/ChatFrame.js

```diff
@@ -562,33 +599,88 @@ class ChatFrame extends React.Component {
-                cancelText="No"
-            ><a href="#"><CloseOutlined style={{color: "red"}}/></a></Popconfirm>)
+        options.push(
+            <div>
+                <Picker set='google' include={['people', 'recent']} onClick={this.sendReaction.bind(this, m)}/>
+            </div>
+        );
+
+
+        options.push(
+            <div>
+                <Popconfirm
+                    key="delete"
+                    title="Are you sure that you want to delete this message?"
+                    onConfirm={this.deleteMessage.bind(this, m)}
+                    okText="Yes"
+                    cancelText="No"
+                >
+                    <a href="#" style={{'color': 'red'}}>Delete the message</a>
+                </Popconfirm>
+            </div>
+
+        )
+
+        const ReactionWrapper = styled.div`
+            height: 20px;
+            overflow: scroll;
+            padding: 0 10px 5px;
+            margin-left: -10px;
+            border-bottom: 1px solid #dcdcdc;
+        `;
+
+        const SingleReactionWrapper = styled.div`
+            float: left;
+            height: 15px;
+            width: auto;
+            margin-left: 8px;
+            border-radius: 10px;
+            .reaction-number {
+                width: 10px;
+                height: 10px;
+                margin-bottom: 2.5px;
+            }   
+        `;
+
         if (options.length > 0)
-            return <Popover key={m.sid} mouseEnterDelay={0.5} placement="topRight" content={<div style={{backgroundColor: "white"}}>
-                {options}
-            </div>}>
+            return <Popover key={m.sid} mouseEnterDelay={0.5} placement="leftTop"
+                            content={<div style={{backgroundColor: "white"}}>
+                                {options}
+                            </div>}>
                 <div ref={(el) => {
-                    this.messagesEnd = el;
-                }} className="chatMessage"><ReactMarkdown source={m.body}
-                                                          renderers={{text: emojiSupport, link: this.linkRenderer}}/>{actionButton}</div>
-            </Popover>
-        return <div key={m.sid} className="chatMessage"><ReactMarkdown source={m.body}
-                                                                       renderers={{
-                                                                           text: emojiSupport,
-                                                                           link: this.linkRenderer
-                                                                       }}/>{actionButton}</div>
+                    this.messagesEnd = el;}} className="chatMessage">
+                    <ReactMarkdown source={m.body} renderers={{text: emojiSupport, link: this.linkRenderer}}/>
+                    {actionButton}
+                </div>
+                {
+                    Object.keys(this.state.reactions).filter(mwe => mwe === m.sid).length === 0 ? null : (
```

Comment:
This is a pretty significant performance anti-pattern: why do an O(n) traversal of the list of keys instead of just access the relevant entry?

Reply:
Yes, this is my mistake. I changed the `reactions` structure but forgot to change here.